src/librustc_mir/transform/inline.rs

```diff
@@ -808,11 +808,19 @@ impl<'a, 'tcx> MutVisitor<'tcx> for Integrator<'a, 'tcx> {
             }
             TerminatorKind::Abort => { }
             TerminatorKind::Unreachable => { }
-            TerminatorKind::FalseEdges { ref mut real_target, ref mut imaginary_targets } => {
+            TerminatorKind::FalseEdges { ref mut real_target, ref mut imaginary_targets,
+                                         ref mut cleanup } => {
                 *real_target = self.update_target(*real_target);
                 for target in imaginary_targets {
                     *target = self.update_target(*target);
                 }
+                if let Some(tgt) = *cleanup {
+                    *cleanup = Some(self.update_target(tgt));
+                } else if !self.in_cleanup_block {
+                    // Unless this assert is in a cleanup block, add an unwind edge to
+                    // the orignal call's cleanup block
+                    *cleanup = self.cleanup_block;
```

Comment:
That said, this doesn't seem right. I feel like if `cleanup` is None, it means that this `FalseEdges` block is not simulating a false unwind, and hence we should not rewrite this edge to anything at all.

Reply:
I agree with your assessment here. Now I'm wondering if it's worthwhile creating a completely separate terminator kind for false unwinding versus false edges since I can see this sort of confusion happening in the future as well. On one hand, it could be useful to express both false edges and false unwinds in one terminator, but on the other we don't use that capability anywhere today and it's already causing problems =)

Reply:
> On one hand, it could be useful to express both false edges and false unwinds in one terminator, but on the other we don't use that capability anywhere today and it's already causing problems

Heh. I don't have a strong opinion here, but a `Terminator::FalseUnwind` variant might feel simpler overall -- I noticed that you rarely seemed to take advantage of the `Option<>`, but instead wound up writing two distinct arms.